### PR TITLE
Improve content negociation middlewares

### PR DESCRIPTION
**Pull request by arteymix**
In short, it handles fuzzy matching and set the appropriate header in the response.

```vala
using Valum.ContentNegotiation;

app.get ("", accept ("text/html", (req, res) => {
    res.body.write_all ("<!DOCTYPE html><html></html>", null);
})).then (accept ("text/xhtml", (req, res) => {
    res.body.write_all ("...");
}, NegociateFlags.FINAL));
```

The `NegociateFlags.FINAL` indicate that this is the last option and a `406 Not Acceptable` should be raised on failure.

It needs some tests before being merged.

### CODE REVIEW

**Comment by arteymix** [comment]
It would be great to have something that allow us to build sequences and connect the last `next` to the keep routing after the sequence.

Something like:

```vala
app.get ("", sequence(
    accept ("text/xml", produce_xml),
    accept ("text/html", produce_html, NegociateFlags.FINAL)
));
```

That could get rid of `then`.

**Comment by arteymix** [comment]
It would be great to handle quality values in some way.

**Comment by arteymix** [comment]
To handle quality values properly, we could provide multiple choices to `negotiate` and let it forward the best possible.

```vala
accept ({"text/html", "text/html+xml"}, (req, res, next, stack, content_type) {
    // ...
}};
```

**Comment by arteymix** [comment]
The tests work, but it produces a SEGFAULT when running the application :(

**Comment by arteymix** [comment]
Okay, everything is correct now! 

Since we handle multiple expectations, the default behaviour is to raise a `406 Not Acceptable`, which can be overwritten by setting the `NegotiateFlags.NEXT`.

No expectations always raise a `406 Not Acceptable` and a missing header forwards the first expectation.

We need some tests for the edge cases and I have some documentation already ;)

**Comment by arteymix** [comment]
The tests have to cover the use of mixed character case, because most of the specification is case-insensitive.

File: src/valum-content-negotiation.vala [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+	 *
+	 * @since 0.3
+	 */
+	public HandlerCallback accept_charset (string[]              charsets,
+	                                       owned ForwardCallback forward = ContentNegotiation.forward,
+	                                       NegotiateFlags        flags   = NegotiateFlags.NONE) {
+		return negotiate ("Accept-Charset", charsets, (req, res, next, stack, charset) => {
+			HashTable<string, string> @params;
+			var content_type = res.headers.get_content_type (out @params);
+			var old_charset  = @params["charset"] ?? "utf-8";
```
Comment (arteymix) [inline src/valum-content-negotiation.vala]: The default charset for HTTP/1.1 is iso-8859-1 (https://www.w3.org/International/O-HTTP-charset), we might want to make that assumption for `accept_charset`. The only thing is, it's not much practical for end-user as app will typically produce UTF-8 data.

**Comment by arteymix** [comment]
The merge will have to deal with a couple of breaking changes:

 - return the result of `forward`
 - stack is a context (not used anyway)

**Comment by arteymix** [comment]
Just picked the codecov commit, so that we could see what parts are missing. It should end-up here: https://codecov.io/github/valum-framework/valum

**Comment by arteymix** [comment]
It would be nice to use a quality list for the expectations and select the best match by taking the highest qvalue product. This is [what Apache uses](http://httpd.apache.org/docs/current/content-negotiation.html).

```vala
app.get ("/", accept ("text/json; q=1, text/xml; q=0.1", (req, res, next, ctx, expectation) => {
    // produce content according to 'expectation'
}));

**Comment by arteymix** [comment]
I merged the changes from the 0.3 so it should be fast-forward into the trunk.

 - support vendor prefix for `accept_encoding`
 - documentation update
 - fix for qvalue extraction
 - remove `NegotiateFlags.NEXT` in favour of a status-handler based approach

Some tests are missing and it should be ready to merge!
